resources/type-system/flow-analysis-v0.md

```diff
@@ -0,0 +1,339 @@
+# Flow Analysis for Non-nullability
+
+paulberry@google.com, leafp@google.com
+
+## Terminology and Notation
+
+We use the generic term *node* to denote an expression, statement, top level
+function, method, constructor, or field.  We assume that all nodes are labelled
+uniquely in some unspecified fashion such that mappings can be maintained from
+nodes to the results of flow analysis.
+
+The flow analysis assumes that the set of variables which are assigned in any
+given node has been pre-computed, and can be queried by a predicate **Assigned**
+such that for a node `N` and a variable `v`, **Assigned**(`N`, `v`) is true iff
+`N` syntactically contains an assignment to `v` (regardless of reachability of
+that assignment).
+
+### Models
+
+A *type test site* is a location in the program where a variable's type is
+tested, either via an `is` expression or a cast.
+
+A *variable model*, denoted `VariableModel(declaredType, promotedTypes,
+testSites, assigned, unassigned, writeCaptured)`, represents what is statically
+known to the flow analysis about the state of a variable at a given point in the
+source code.
+
+- `declaredType` is the type assigned to the variable at its declaration site
+  (either explicitly or by type inference).
+
+- `promotedTypes` is an ordered set of types that the variable has been promoted
+  to, with the final entry in the ordered set being the current promoted type of
+  the variable.  Note that each entry in the ordered set must be a subtype of
+  all previous entries, and of the declared type.
+
+- `testSites` is an ordered list of *type test site*s representing the set of
+  type tests that are known to have been performed on the variable in all code
+  paths leading to the given point in the source code.
+
+- `assigned` is a boolean value indicating whether the variable is known to
+  have been definitely assigned at the given point in the source code.
+
+- `unassigned` is a boolean value indicating whether the variable is known not
+  to have been definitely assigned at the given point in the source code.  (Note
+  that a variable may not be both definitely assigned and definitely
+  unassigned).
+
+- `writeCaptured` is a boolean value indicating whether a closure might exist at
```

Comment:
I'm starting to implement the logic for `writeCaptured` in the prototype, and I wanted to let you know about two corner cases I've run into.  First, what if a write happens in one closure and a read happens in another:
```dart
f(Object x) {
  void Function() h;
  var g = () {
    if (x is int) { // (1)
      h();
      x.isEven; // (2)
    }
  };
  h = () { x = "hello"; }; // (3)
  g();
}
```
Even though the if-test at (1) appears lexically before the closure that captures a write to `x` is created (at (3)), since (1) is itself inside a closure, there's no guarantee that it will execute prior to the write being captured.  So promotion is unsafe in this case, and (2) should be a compile-time error.  I'm handling this situation by saying that prior to analysis, we gather a list of all variables that are written to in all closures.  Then, when entering any closure, analysis sets `writeCaptured` to true for any variables in that list.

Second, what if a write capturing closure happens in a loop, after a promotion attempt?
```dart
f(Object x) {
  void Function()? g;
  while (true) {
    if (x is int) { // (1)
      if (g != null) g();
      x.isEven; // (2)
    }
    g = () { x = 'hello'; };
  }
}
```
In this case the promotion at (1) is unsafe, because the second time through the loop, there is a write captured.  So there should be a compile-time error at (2).  I'm handling this situation by saying that when entering a loop, we set `writeCaptured` to true for any variables that are captured ahywhere within the loop.